### ISSUE TITLE
scale shards count issue

### ISSUE THREAD

**Issue by Kylinrix** [Altinity/clickhouse-operator#727]
There is a issue when we change the shard count.
It is a cluster (named ck100032773) has local tables, Distributed engine tables, and some other momery engine table.
ck100032773 cluster has 1 shard and 2 replicas, There are sts name:
```
chi-ck100032773-ck100032773-0-0
chi-ck100032773-ck100032773-0-1
```
The FQDN are
```
chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local 
chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local
```
when we changed the shard count from 1 to 2, The operation will create `chi-ck100032773-ck100032773-1-0` and `chi-ck100032773-ck100032773-1-1` sts and create all existing schema in the right situation.

Log shows the operator call `HostCreateTables`,`getCreateReplicaObjects`,`getCreateReplicaObjectsContext` functions to get all schema Query.
But It seems operator cannot get The right connection when call "getCreateReplicaObjectsContext" function.

**But We find the `getCreateReplicaObjectsContext` function make the `chi-ck100032773-ck100032773-1-1` host to be its host arg all the time.** 

**This operation causes the remote url in sql is always `chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local` which is 1-1's FQDN.**
We think that 1-1's service or connection is not ready when call the SQL Query

`getCreateReplicaObjects` function is as follows:
```go
func (s *Schemer) getCreateReplicaObjects(host *chop.ChiHost) ([]string, []string, error) {

	var shard *chop.ChiShard = nil
	var replicaIndex int
	for shardIndex := range host.GetCluster().Layout.Shards {  //replica is host 
		shard = &host.GetCluster().Layout.Shards[shardIndex]
		for replicaIndex = range shard.Hosts {
			replica := shard.Hosts[replicaIndex]
			if replica == host {
				break
			}
		}
	}

        ...

	replicas := CreatePodFQDNsOfShard(shard) // only get new shard's host
	nReplicas := len(replicas)
	if nReplicas <= 1 {
		log.V(1).M(host).F().Info("Single replica in a shard. Nothing to create a schema from.")
		return nil, nil, nil
	}

	replicas[replicaIndex] = replicas[nReplicas-1]
	replicas = replicas[:nReplicas-1]
	log.V(1).M(host).F().Info("Extracting replicated table definitions from %v", replicas)
       
        //The  replicas is always new shard's host,.
        //It always be 'chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local' that has no schema in our case
	system_tables := fmt.Sprintf("remote('%s', system, tables)", strings.Join(replicas, ",")) 
       
        ...

```

How to fix it ?

version: Clickhouse Operator v.0.13.5

The operator log is as follows

```
I0611 04:00:56.992064       1 announcer.go:213] getCreateReplicaObjects():uat-clickhouse/ck100032773:Extracting replicated table definitions from [chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
I0611 04:00:56.992213       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.026097       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local:8123/) clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
 for SQL: SELECT DISTINCT
        database AS name,
        concat('CREATE DATABASE IF NOT EXISTS "', name, '"') AS create_db_query
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.026134       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
I0611 04:00:57.026156       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.062128       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local:8123/) clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
 for SQL: SELECT DISTINCT
        database AS name,
        concat('CREATE DATABASE IF NOT EXISTS "', name, '"') AS create_db_query
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.062166       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
I0611 04:00:57.062183       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.089670       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local:8123/) clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
 for SQL: SELECT DISTINCT
        database AS name,
        concat('CREATE DATABASE IF NOT EXISTS "', name, '"') AS create_db_query
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.089725       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
I0611 04:00:57.089744       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.093096       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local:8123/) dial tcp: lookup chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local on 10.7.0.10:53: no such host for SQL: SELECT DISTINCT
        database AS name,
        concat('CREATE DATABASE IF NOT EXISTS "', name, '"') AS create_db_query
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.093130       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: dial tcp: lookup chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local on 10.7.0.10:53: no such host
E0611 04:00:57.093170       1 announcer.go:260] schemer.go:84:getObjectListFromClickHouse():FAILED to run query on all endpoints [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
I0611 04:00:57.093299       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.114876       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local:8123/) clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
 for SQL: SELECT DISTINCT
        name,
        replaceRegexpOne(create_table_query, 'CREATE (TABLE|VIEW|MATERIALIZED VIEW|DICTIONARY)', 'CREATE \\1 IF NOT EXISTS')
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system' and create_table_query != '' and name not like '.inner.%'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.114931       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
I0611 04:00:57.114980       1 announcer.go:213] Run query on: chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local]
E0611 04:00:57.137059       1 announcer.go:262] connection.go:91:Query():FAILED Query(http://***:***@chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local:8123/) clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


 (version 20.8.18.32 (official build))
 for SQL: SELECT DISTINCT
        name,
        replaceRegexpOne(create_table_query, 'CREATE (TABLE|VIEW|MATERIALIZED VIEW|DICTIONARY)', 'CREATE \\1 IF NOT EXISTS')
FROM remote('chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local', system, tables)
WHERE database != 'system' and create_table_query != '' and name not like '.inner.%'
SETTINGS skip_unavailable_shards = 1
W0611 04:00:57.137103       1 announcer.go:240] schemer.go:80:getObjectListFromClickHouse():FAILED to run query on: chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local of [chi-ck100032773-ck100032773-0-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-0-1.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-0.uat-clickhouse.svc.cluster.local chi-ck100032773-ck100032773-1-1.uat-clickhouse.svc.cluster.local] skip to next. err: clickhouse: Code: 519, e.displayText() = DB::NetException: All attempts to get table structure failed. Log:


```

**Comment by alex-zaitsev** [comment]
@Kylinrix, the logic in 0.13.5 has some shortages, and we've made improvements since then. Could you try with 0.14.0 or even 0.15.0 (not yet released, but image is available).

**Comment by alex-zaitsev** [comment]
Fixed in 0.15.0
